### ISSUE TITLE
Keyboard mapping: German keyboard special characters wrong (server: Win10, client:OpenSuse15.1, inverse to #1665)

### ISSUE THREAD

**Issue by FredFirestone** [Xpra-org/xpra#3029]
I'm using xpra to shadow a windows 10 desktop to a linux client. Unfortunately german umlaut keys aren't working - they are as if not pressed at all.
However debugging on the server indicates keys are actually (properly) received (e.g. ö as o-diaeresis):
```
2021-02-20 20:48:21,953 get_keyboard_info took 0ms
2021-02-20 20:48:40,184 get_keycode(47, 'odiaeresis', True, [], 0)=-1
2021-02-20 20:48:40,184 process_key_action(['key-action', 1, b'odiaeresis', True, (), 246, b'\xc3\xb6', 47, 0]) server keycode=-1, group=0
2021-02-20 20:48:40,184 make_keymask_match([], -1, ['odiaeresis'])
2021-02-20 20:48:40,194 keys pressed=
2021-02-20 20:48:40,194 make_keymask_match: current mask=set(), wanted=set(), ignoring=-1/['odiaeresis']
2021-02-20 20:48:40,334 process_key_action(['key-action', 1, b'odiaeresis', False, (), 246, b'\xc3\xb6', 47, 0]) server keycode=-1, group=0
2021-02-20 20:48:40,334 make_keymask_match([], -1, ['odiaeresis'])
2021-02-20 20:48:40,344 keys pressed=
2021-02-20 20:48:40,344 make_keymask_match: current mask=set(), wanted=set(), ignoring=-1/['odiaeresis']
2021-02-20 20:48:43,094 get_keycode(48, 'adiaeresis', True, [], 0)=-1
2021-02-20 20:48:43,094 process_key_action(['key-action', 1, b'adiaeresis', True, (), 228, b'\xc3\xa4', 48, 0]) server keycode=-1, group=0
2021-02-20 20:48:43,104 make_keymask_match([], -1, ['adiaeresis'])
2021-02-20 20:48:43,104 keys pressed=
2021-02-20 20:48:43,104 make_keymask_match: current mask=set(), wanted=set(), ignoring=-1/['adiaeresis']
2021-02-20 20:48:43,194 process_key_action(['key-action', 1, b'adiaeresis', False, (), 228, b'\xc3\xa4', 48, 0]) server keycode=-1, group=0
2021-02-20 20:48:43,194 make_keymask_match([], -1, ['adiaeresis'])
2021-02-20 20:48:43,194 keys pressed=
2021-02-20 20:48:43,204 make_keymask_match: current mask=set(), wanted=set(), ignoring=-1/['adiaeresis']
2021-02-20 20:54:11,391 get_keycode(37, 'Control_L', True, [], 0)=162
2021-02-20 20:54:11,401 process_key_action(['key-action', 1, b'Control_L', True, (), 65507, b'', 37, 0]) server keycode=162, group=0
2021-02-20 20:54:11,401 make_keymask_match([], 162, ['Control_L'])
2021-02-20 20:54:11,401 keys pressed=
2021-02-20 20:54:11,411 make_keymask_match: current mask=set(), wanted=set(), ignoring=162/['Control_L']
2021-02-20 20:54:11,411 handle_key((1, True, 'Control_L', 65507, 162, [], True, True))
2021-02-20 20:54:11,411 handle keycode pressing   162: key 'Control_L'
2021-02-20 20:54:11,411 fake_key(162, True) calling keybd_event(162, 29, 0, 0)
2021-02-20 20:54:11,581 get_keycode(54, 'c', True, ['control'], 0)=67
2021-02-20 20:54:11,581 process_key_action(['key-action', 1, b'c', True, (b'control',), 99, b'\x03', 54, 0]) server keycode=67, group=0
2021-02-20 20:54:11,591 make_keymask_match(['control'], 67, ['c'])
2021-02-20 20:54:11,591 keys pressed=
2021-02-20 20:54:11,591 make_keymask_match: current mask={'control'}, wanted={'control'}, ignoring=67/['c']
2021-02-20 20:54:11,591 handle_key((1, True, 'c', 99, 67, ['control'], False, True))
2021-02-20 20:54:11,601 handle keycode pressing    67: key 'c'
2021-02-20 20:54:11,601 fake_key(67, True) calling keybd_event(67, 46, 0, 0)
2021-02-20 20:54:11,601 scheduling key repeat timer with delay 600 for c / 67
```
So I'm a bit at a loss, why the key's aren't properly showing up in application windows...
[BugReport.zip](https://github.com/Xpra-org/xpra/files/6016263/BugReport.zip)

**Comment by totaam** [comment]
That's because we don't find the key: `get_keycode(48, 'adiaeresis', True, [], 0)=-1`
I assume that the layout you're using is this one:
![german keyboard layout](https://upload.wikimedia.org/wikipedia/commons/thumb/3/36/KB_Germany.svg/800px-KB_Germany.svg.png)
And that you're pressing the key to the right of `L`?

If that's the case then this change works for me: d5cf1aa5e5c8a097588c2dc12c6c56ff69cfffdc
In fact, it's a much more reliable way of doing the key mapping on MS Windows.
I have uploaded a beta build with this fix here: https://xpra.org/beta/windows/ (r28487 or later)

If that works for you, please close this ticket.
If not, is the Win10 server configured with a german keyboard layout as main layout?
Can you post the output of the `Keyboard_info.exe` tool?

**Comment by FredFirestone** [comment]
The keyboard used is like the picture, the key is exactly as assumed (right to L). Some others (Ä,Ü) aren't working, but 'ß' does work.

Xpra version was Xpra 3.0.12 (64-bit, actually 3.0.12-r27627), I assume it was python3 but don't know for sure.

Keyboard_Info.exe returns

> Modifiers:
> 
> Server Managed                    : None
> Missing from pointer events       : lock
> 
> keyboard layout code 0x407
> identified as 'German' : de
> Layout:     'de'
> Layouts:    'de', 'cz'
> Variant:    ''
> Variants:   'nodeadkeys', 'sundeadkeys', 'mac'
> Options:
> 
> Repeat:     500, 33

The server has german keyboard as default (but windows option to switch layout is enabled). 

I've yet to try the build, thank you very much for the quick response.

**Comment by FredFirestone** [comment]
I've tried Xpra-Python3-x86_64_Setup_4.1-r28487.exe and it did fix the umlaut problems, thank you very much.
There are now much fewer keys not working. I've found not working
- all blue keys in above graphic (entered together with "Alt-Gr"-Key) 
(most or even all seem to get transmitted, debug shows "twosuperior" for ², "bar" for |, asciitilde for ~, ...:); debug log example:
make_keymask_match: current mask=set(), wanted={'mod5'}, ignoring=187/['asciitilde']
- keys with dead accent (not relevant in german and thus I very rarely use it)
As the original request is completely satisfied, the issue could be closed. If the blue keys could be made working as well, it would certainly great.
